### PR TITLE
Integrate conditional loss

### PR DESCRIPTION
**Pull request by davidleonfdez**
This PR integrates the conditional loss in the training notebook and tests it.

### CODE REVIEW

**Comment by davidleonfdez** [comment]
It's not finished yet.

I've had some setbacks:

- The `GPT2ForSequenceClassification` isn't able to handle padded inputs when it receives `inputs_embeds` instead of `input_ids`. I opened an issue suggesting a workaround in the transformers repository.
- It's hard to fit the models in the memory of a single GPU. We need to have 3 GPT2 in GPU memory at the same time (or at least 2 if we engineer some kind of alternate updates strategy that requires moving one classifier from GPU to CPU and the other from CPU to GPU, not worth it for now)
  - The first solution I tried (last commit) is not enough.
  - Next, I'll try training with mixed precision (fp16) and/or gradient checkpointing

cc @ismaelfaro

**Comment by davidleonfdez** [comment]
I could finally solve the OOM issues. Although I did some minor changes I've changed my mind and now I doubt that there was any memory leak because after reducing a bit the memory needs this conditional model is able to train as many epochs as you want.

From what I've researched the allocated GPU memory might increase gradually during an epoch depending on the size of the inputs and the order of the examples, how they are processed, ..., giving the false impression of a memory leak.

**Comment by davidleonfdez** [comment]
Talking about the results, the conditional loss makes the regular validation loss (column "Best metadata-less val. loss") get a bit worse, but the conditional metrics improve a lot, in some cases above my expectations.

The reasons may be diverse:
- The conditional loss has made the generator learn to produce text close to the mean of the distribution of the poems of its category, which can mean less variety.
- Overfitting could explain it too, because if the generator produces sequences more similar to the training examples, they should be easier to classify because the classifier used for evaluation was trained with the same train/validation splits. However, the training loss at the end of training (early stopping actually) is similar to the rest of the configurations tested.
- Although the classifiers used in the conditional loss and the classifiers used for evaluation are different, they can share some common bias because they were trained with the same data. So, the mistakes that make the conditional loss classifiers influence poorly the generator could be mistakes that the evaluation classifiers are prone to make too while assigning a category to a poem of the validation set.
- This loss is awesome haha
- Some combination of the above.

I'm leaning toward the last option.

We'll have to test the generation to get a feeling of the quality of the outputs, but I think this is ready to merge @ismaelfaro.
